### PR TITLE
CoW: Add reference tracking to index when created from series

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel this takes a shot at #34364 under CoW. Still a rough poc that is missing tests for the other index classes, but would appreciate if you could take a quick look. This avoids modifying an index by modifying the object it was created from.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -477,6 +482,10 @@ def __new__(
 
         data_dtype = getattr(data, "dtype", None)
 
+        refs = None
+        if not copy and isinstance(data, (ABCSeries, Index)):
+            refs = data._references
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/indexes/base.py:487]: Yep exactly. But I wanted to wait for the pr that tackles this for the DataFrame case to get merged before adding this for index

**Comment by phofl** [comment]
Yes exactly, we are doing a copy in this case right now if I am not mistaken, so I kept it out of here on purpose. But this is certainly a sensible follow up

**Comment by phofl** [comment]
So you'd say this is generally ok to merge?

**Comment by jorisvandenbossche** [comment]
For me, yes

**Comment by phofl** [comment]
Rebasing once more

**Comment by phofl** [comment]
merging to get started with the follow ups

File: pandas/_libs/internals.pyx [hunk 2]
```diff
@@ -890,6 +890,16 @@ cdef class BlockValuesRefs:
         """
         self.referenced_blocks.append(weakref.ref(blk))
 
+    def add_index_reference(self, index: object) -> None:
+        """Adds a new reference to our reference collection when creating an index.
+
+        Parameters
+        ----------
+        index: object
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/_libs/internals.pyx:898]: Yes

**Pull request merged**